docs/broker/filtering.md

```diff
@@ -0,0 +1,618 @@
+
+The CEL language is new and has limited exposure, despite being informed by
+experience and research at Google. We may discover issues that make it less
+suitable for Trigger filtering (e.g.
+[google/cel-go#203](https://github.com/google/cel-go/issues/203).
+
+If CEL turns out to be a liability, a new expression language could be added
+alongside CEL without breaking existing triggers, allowing for an orderly
+deprecation.
+
+For example, a javascript filter could be added without disrupting existing CEL
+filters:
+
+```yaml
+spec:
+  filter:
+    javascript:
+      expression: ce.type == "com.github.pull.create"
+```
+
+### Expression language versioning
+
+Embedding an expression language into the Trigger specification means that we
+effectively have two versions exposed to the user: the Trigger CRD APIVersion and
+the expression language version. As the CEL spec evolves, backward incompatible
+changes may be made that would require a version change in the Trigger.
+
+I propose that we rely on the Trigger CRD APIVersion to encode both the Trigger
+version and the CEL version. This ensures that Triggers with the same APIVersion
+always have the same behavior, and the upgrade semantics are the same as those
+for other Kubernetes objects.
+
+There is a concern that tying the CEL version to the Trigger APIVersion could
+block users from upgrading, because they need to upgrade all of their
+expressions at the same time as a Knative upgrade. If no automatic upgrade is
+possible, we should support both CEL versions for one or more Knative releases.
+It might be necessary to add a Trigger field specifying the version of CEL that
+should evaluate the expression.
+
+### Numeric and nested extension values
+
+Some CloudEvents transports (such as the HTTP binary content mode) may be unable
+to differentiate numeric and string values for top-level or nested extension
+attributes. This is still being discussed by the CloudEvents working group (see
+https://github.com/cloudevents/spec/pull/413).
+
+Until this is resolved, the type of numeric values parsed in CloudEvents
+extensions may be unpredictable.
+
+### Single top-level filter
+
+Top-level filters are now explicitly exclusive: only one may be specified. This
+is done to ensure that Trigger filters are easily understood by the user.
+Composable filters at the top-level would have unclear composition logic: are
+they composed by Boolean **OR** or **AND**? The user must be aware of the
+implicit semantics to understand the result of the trigger. 
+
+If there is user demand for structured filter composition, top-level `or` and
+`and` filters can be introduced that indicate how filter elements are composed.
+This example is presented as an illustration only; it is not an element of the
+proposal.
+
+```yaml
+spec:
+  filter:
+    and:
+    - attributes:
+        type: com.github.pull.create
+    - expression: ce.source.match("knative")
+```
+
+## Alternatives Considered
```

Comment:
I do not see the alternative of `just a function`. Did you consider implementing a service service wrapper that implements CEL parser and then use that as a ref in the filter spec? It would use the same mechanize as the subscriptions.

Reply:
For example, use this entire proposal but make a new trigger type? 

TriggerCEL -makes-> a serving.service + a trigger with a function ref. And then promote the CEL syntax up to to the spec level?

Reply:
I think this would be a useful addition to this section, but I'm a little unclear what UI is being proposed. Can you give an example of the API? Can you briefly describe the benefits and disadvantages from the user's perspective?